rust/agama-server/src/network/web.rs

```diff
@@ -167,67 +175,142 @@ async fn connections(State(state): State<NetworkState>) -> Json<Vec<NetworkConne
         .unwrap();
     let _ = rx.await.unwrap();
 
-    let conn = Connection::try_from(conn)?;
-    let uuid = conn.uuid.clone();
-
+    let (tx, rx) = oneshot::channel();
     state
         .actions
-        .send(Action::UpdateConnection(Box::new(conn)))
+        .send(Action::GetConnection(id.clone(), tx))
         .unwrap();
 
-    Ok(Json(uuid))
+    match rx.await.unwrap() {
+        None => Err(NetworkError::CannotAddConnection(id.clone())),
+        Some(conn) => Ok(Json(conn)),
+    }
 }
 
-#[utoipa::path(delete, path = "/network/connections/:uuid", responses(
+#[utoipa::path(delete, path = "/network/connections/:id", responses(
   (status = 200, description = "Delete connection", body = Connection)
 ))]
 async fn delete_connection(
     State(state): State<NetworkState>,
-    Path(id): Path<Uuid>,
+    Path(id): Path<String>,
 ) -> Result<Json<()>, NetworkError> {
     let (tx, rx) = oneshot::channel();
     state
         .actions
         .send(Action::RemoveConnection(id, tx))
         .unwrap();
 
-    let _ = rx.await.unwrap();
-    let (tx, rx) = oneshot::channel();
-    state.actions.send(Action::Apply(tx)).unwrap();
-    let _ = rx.await.unwrap();
-
-    Ok(Json(()))
+    Ok(Json(rx.await.unwrap()?))
 }
 
-#[utoipa::path(put, path = "/network/connections/:uuid", responses(
+#[utoipa::path(put, path = "/network/connections/:id", responses(
   (status = 200, description = "Update connection", body = Connection)
 ))]
 async fn update_connection(
     State(state): State<NetworkState>,
-    Path(id): Path<Uuid>,
+    Path(id): Path<String>,
     Json(conn): Json<NetworkConnection>,
 ) -> Result<Json<()>, NetworkError> {
+    let (tx, rx) = oneshot::channel();
+    state
+        .actions
+        .send(Action::GetConnection(id.clone(), tx))
+        .unwrap();
+    let current_conn = rx.await.unwrap();
     let mut conn = Connection::try_from(conn)?;
-    conn.uuid = id;
+    match current_conn {
+        Some(current) => {
+            if current.id != id.clone() {
+                return Err(NetworkError::UnknownConnection(id));
+            } else {
+                conn.uuid = current.uuid;
+            }
+        }
+        None => return Err(NetworkError::UnknownConnection(id)),
+    }
+
+    let (tx, rx) = oneshot::channel();
+    state
+        .actions
+        .send(Action::UpdateConnection(Box::new(conn), tx))
+        .unwrap();
 
+    Ok(Json(rx.await.unwrap()?))
+}
+
+#[utoipa::path(get, path = "/network/connections/:id/connect", responses(
+  (status = 200, description = "Connect to the given connection", body = String)
+))]
+async fn connect(
+    State(state): State<NetworkState>,
+    Path(id): Path<String>,
+) -> Result<Json<()>, NetworkError> {
+    let (tx, rx) = oneshot::channel();
+    state
+        .actions
+        .send(Action::GetConnection(id.clone(), tx))
+        .unwrap();
+
+    if let Some(mut current_conn) = rx.await.unwrap() {
+        current_conn.set_up();
```

Comment:
Would not be better to put this logic as part of the action? IMHO, as much as possible, the handler should be responsible of parsing/validating the request/response and call the right method (or dispatch the right action).

However, if we have followed that approach elsewhere, we can keep it as it is and reevaluate in the short term (when we drop the D-Bus code and clean-up the actions).